src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java

```diff
@@ -2111,7 +2111,7 @@ public boolean overrides(Symbol _other, TypeSymbol origin, Types types, boolean
 
             // check for a direct implementation
             if (other.isOverridableIn((TypeSymbol)owner) &&
-                types.asSuper(owner.type, other.owner) != null) {
+                types.asSuper(owner.type.referenceProjectionOrSelf(), other.owner) != null) {
```

Comment:
So, for source ovrride we use `asSuper`, but for binary override we use `isSubtype`. Shouldn't we use asSuper in both cases here? Overriding is typically a business between class declarations, so symbols should be good? Or can you envision a case where you want to check if a method on Foo.val overrides something on a Foo.ref and you want to say "no" there? In other words, I think the check here is meant to say: make sure that the owner of the method symbols are related by subclassing - which seems to call for asSuper.

Reply:
Agreed.